Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp

```diff
@@ -143,17 +148,86 @@ static bool canPerformAcceleratedRendering()
     return ProcessCapabilities::canUseAcceleratedBuffers() && PlatformDisplay::sharedDisplay().skiaGLContext();
 }
 
-RenderingMode SkiaPaintingEngine::renderingMode() const
+bool SkiaPaintingEngine::shouldUseGPURenderingForDirtyRect(const IntRect& dirtyRect) const
+{
+    // Single strategy: If CPU is idle, always use it.
+    auto handlePreferCPUIfIdle = [&]() -> bool {
+        return m_cpuWorkerPool && !m_cpuWorkerPool->numberOfTasks();
+    };
+
+    // Single strategy: If GPU is idle, always use it.
+    auto handlePreferGPUIfIdle = [&]() -> bool {
+        return m_gpuWorkerPool && !m_gpuWorkerPool->numberOfTasks();
+    };
+
+    // Single strategy: If painting area exceeds a threshold, always use GPU.
+    auto handlePreferGPUAboveMinimumArea = [&]() -> bool {
+        return dirtyRect.area() >= minimumAreaForGPUPainting();
+    };
+
+    // Single strategy: Decide randomly whether to use GPU or not.
+    auto handleMinimumFractionOfTasksUsingGPU = [&]() -> bool {
+        auto randomFraction = static_cast<double>(weakRandomNumber<uint32_t>()) / static_cast<double>(UINT32_MAX);
+        ASSERT(randomFraction >= 0);
+        ASSERT(randomFraction <= 1);
+        return randomFraction <= minimumFractionOfTasksUsingGPUPainting();
+    };
+
+    // Combined strategy: default for WPE, "hybrid mode", saturates CPU painting, before using GPU.
+    auto handleCPUAffineRendering = [&]() -> bool {
+        // If the CPU worker pool has unused workers, use them.
+        if (m_cpuWorkerPool && m_cpuWorkerPool->numberOfTasks() < numberOfCPUPaintingThreads())
+            return false;
+
+        // If the GPU worker pool has unused workers, use them.
+        if (m_gpuWorkerPool && m_gpuWorkerPool->numberOfTasks() < numberOfGPUPaintingThreads())
+            return true;
+
+        return handleMinimumFractionOfTasksUsingGPU();
+    };
+
+    // Combined strategy: default for Gtk, useful mode for high-end GPUs, saturates GPU painting, before using CPU.
+    auto handleGPUAffineRendering = [&]() -> bool {
+        // If the GPU worker pool has unused workers, use them.
+        if (m_gpuWorkerPool && m_gpuWorkerPool->numberOfTasks() < numberOfGPUPaintingThreads())
+            return false;
```

Comment:
Shouldn't this return true?

Reply:
yes, there's a mistake in the logic, I'll revise it.